src/jdk.naming.dns/share/classes/com/sun/jndi/dns/DnsClient.java

```diff
@@ -223,20 +227,27 @@ ResourceRecords query(DnsName fqdn, int qclass, int qtype,
 
         Exception caughtException = null;
         boolean[] doNotRetry = new boolean[servers.length];
-
+        // Holder for unfulfilled timeouts left for each server
+        AtomicLong[] unfulfilledUdpTimeouts = IntStream.range(0, servers.length)
+                .mapToObj(_ -> new AtomicLong())
+                .toArray(AtomicLong[]::new);
         try {
             //
             // The UDP retry strategy is to try the 1st server, and then
             // each server in order. If no answer, double the timeout
             // and try each server again.
             //
-            for (int retry = 0; retry < retries; retry++) {
-
+            for (int retry = 0; retry <= retries; retry++) {
+                boolean isLastRetry = retry == retries;
                 // Try each name server.
                 for (int i = 0; i < servers.length; i++) {
                     if (doNotRetry[i]) {
                         continue;
                     }
+                    AtomicLong unfulfilledServerTimeout = unfulfilledUdpTimeouts[i];
```

Comment:
```suggestion
                    // unfulfilledServerTimeout is always >= 0
                    AtomicLong unfulfilledServerTimeout = unfulfilledUdpTimeouts[i];
```

Reply:
Thanks for the suggestion. Added in [793cfda](https://github.com/openjdk/jdk/pull/20892/commits/793cfda3282788d5af5efa57db6eeb58641eb3ed)